{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c45c37d-e183-4674-9b8d-c5e1146043c4",
   "metadata": {},
   "source": [
    "could you please please make this into a text with a chapter for the repo gh \"#!/usr/bin/env python3\n",
    "\"\"\"\n",
    "Scans 1000 Genomes chr19 VCF for variants overlapping APOE/TOMM40 3'UTRs and\n",
    "predicts miRNA seed gain/loss by comparing reference and alternate sequences.\n",
    "\n",
    "Requirements:\n",
    "    pip install cyvcf2 biopython pandas intervaltree tqdm\n",
    "Run:\n",
    "    python apoe_mirna_scan.py\n",
    "\n",
    "Outputs:\n",
    "    results_apoe_miRNA_seed_gain_loss.csv\n",
    "\"\"\"\n",
    "\n",
    "\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f43c2a4d-29c8-4868-9352-0a915041534a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from cyvcf2 import VCF\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "from intervaltree import IntervalTree\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f1ed586d-f2ea-4884-910c-4cd43cfe6a86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing GTF for 3'UTR coordinates (genes: APOE, TOMM40)...\n",
      "##description: evidence-based annotation of the human genome (GRCh37), version 19 (Ensembl 74)\n",
      "\n",
      "##provider: GENCODE\n",
      "\n",
      "##contact: gencode@sanger.ac.uk\n",
      "\n",
      "##format: gtf\n",
      "\n",
      "##date: 2013-12-05\n",
      "\n",
      "  APOE: 48 3'UTR segments\n",
      "  TOMM40: 134 3'UTR segments\n",
      "Loading chr19 reference FASTA...\n",
      "Reference length: 59128983\n",
      "Loading miRNA seeds from miRBase mature.fa ...\n",
      "Loaded 2656 human miRNA seeds\n",
      "Scanning VCF for variants overlapping 3'UTRs...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1832506it [02:17, 13315.60it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total candidate variant rows: 578\n",
      "Wrote results_apoe_miRNA_seed_gain_loss.csv\n",
      "\n",
      "Top candidate variants (top 20 by predicted seed gain/loss):\n",
      "CHROM      POS REF ALT   GENE  n_GAIN  n_LOSS   AF                                                                                                                                                                                           GAIN_miRNAs                                                                                                                                                                                                               LOSS_miRNAs\n",
      "   19 45411748   C   T   APOE      12       9 None                           hsa-miR-1207-5p|hsa-miR-1827|hsa-miR-3612|hsa-miR-4270|hsa-miR-4441|hsa-miR-4763-3p|hsa-miR-650|hsa-miR-6754-5p|hsa-miR-6760-5p|hsa-miR-6808-5p|hsa-miR-6893-5p|hsa-miR-940                                                                                  hsa-miR-10396b-5p|hsa-miR-1908-5p|hsa-miR-3175|hsa-miR-4706|hsa-miR-4749-5p|hsa-miR-663a|hsa-miR-6777-5p|hsa-miR-6787-5p|hsa-miR-6889-5p\n",
      "   19 45410250   G   C   APOE      14       7 None hsa-miR-11181-3p|hsa-miR-1321|hsa-miR-3150b-3p|hsa-miR-4510|hsa-miR-4739|hsa-miR-4756-5p|hsa-miR-4784|hsa-miR-6127|hsa-miR-6129|hsa-miR-6130|hsa-miR-6133|hsa-miR-6834-5p|hsa-miR-7106-5p|hsa-miR-765                                                                                                                  hsa-miR-1827|hsa-miR-4649-3p|hsa-miR-4722-5p|hsa-miR-4768-3p|hsa-miR-6808-5p|hsa-miR-6893-5p|hsa-miR-940\n",
      "   19 45402516   G   A TOMM40      13       6 None                  hsa-let-7a-5p|hsa-let-7b-5p|hsa-let-7c-5p|hsa-let-7d-5p|hsa-let-7e-5p|hsa-let-7f-5p|hsa-let-7g-5p|hsa-let-7i-5p|hsa-miR-4266|hsa-miR-4458|hsa-miR-4500|hsa-miR-7162-3p|hsa-miR-98-5p                                                                                                                                  hsa-miR-1827|hsa-miR-4649-3p|hsa-miR-4722-5p|hsa-miR-6808-5p|hsa-miR-6893-5p|hsa-miR-940\n",
      "   19 45409089   C   T   APOE       5      12 None                                                                                                                                   hsa-miR-1200|hsa-miR-3915|hsa-miR-4476|hsa-miR-544b|hsa-miR-6876-5p                                   hsa-miR-1343-3p|hsa-miR-2110|hsa-miR-4271|hsa-miR-4667-5p|hsa-miR-4700-5p|hsa-miR-4725-3p|hsa-miR-4747-5p|hsa-miR-5196-5p|hsa-miR-6780b-5p|hsa-miR-6783-3p|hsa-miR-6852-5p|hsa-miR-8089\n",
      "   19 45394544   C   T TOMM40      11       5 None                                                     hsa-miR-3116|hsa-miR-3690|hsa-miR-378a-3p|hsa-miR-378b|hsa-miR-378c|hsa-miR-378d|hsa-miR-378e|hsa-miR-378f|hsa-miR-378h|hsa-miR-378i|hsa-miR-422a                                                                                                                                                     hsa-miR-10400-3p|hsa-miR-378g|hsa-miR-4674|hsa-miR-661|hsa-miR-939-3p\n",
      "   19 45403222   C   G TOMM40       0      15 None                                                                                                                                                                                                       hsa-miR-1321|hsa-miR-149-3p|hsa-miR-4510|hsa-miR-4728-5p|hsa-miR-4739|hsa-miR-4756-5p|hsa-miR-6127|hsa-miR-6129|hsa-miR-6130|hsa-miR-6133|hsa-miR-6758-5p|hsa-miR-6785-5p|hsa-miR-6856-5p|hsa-miR-6873-5p|hsa-miR-6883-5p\n",
      "   19 45411263   C   T   APOE       4      11 None                                                                                                                                             hsa-miR-1321|hsa-miR-3162-5p|hsa-miR-4739|hsa-miR-4756-5p                                                        hsa-miR-3150a-3p|hsa-miR-3661|hsa-miR-4667-5p|hsa-miR-4700-5p|hsa-miR-492|hsa-miR-631|hsa-miR-6763-5p|hsa-miR-6825-5p|hsa-miR-6852-5p|hsa-miR-7851-3p|hsa-miR-8089\n",
      "   19 45396973   G   A TOMM40      13       1 None          hsa-miR-1587|hsa-miR-3620-5p|hsa-miR-4492|hsa-miR-4498|hsa-miR-4505|hsa-miR-4731-5p|hsa-miR-5001-5p|hsa-miR-5787|hsa-miR-6737-5p|hsa-miR-6812-5p|hsa-miR-6819-5p|hsa-miR-7155-5p|hsa-miR-762                                                                                                                                                                                                           hsa-miR-6781-5p\n",
      "   19 45402090   C   T TOMM40       4       9 None                                                                                                                                            hsa-miR-3622a-5p|hsa-miR-3975|hsa-miR-4423-3p|hsa-miR-6069                                                                                         hsa-miR-1285-3p|hsa-miR-3187-5p|hsa-miR-4505|hsa-miR-485-5p|hsa-miR-5189-5p|hsa-miR-5787|hsa-miR-612|hsa-miR-6860|hsa-miR-6884-5p\n",
      "   19 45400727   G   A TOMM40       4       8 None                                                                                                                                              hsa-miR-197-5p|hsa-miR-3132|hsa-miR-4451|hsa-miR-6755-5p                                                                                             hsa-miR-2467-3p|hsa-miR-3127-5p|hsa-miR-4257|hsa-miR-4436b-3p|hsa-miR-4632-5p|hsa-miR-6735-5p|hsa-miR-6879-5p|hsa-miR-7843-5p\n",
      "   19 45403458   C   T TOMM40      12       0 None            hsa-miR-302a-3p|hsa-miR-302b-3p|hsa-miR-302c-3p|hsa-miR-302d-3p|hsa-miR-302e|hsa-miR-372-3p|hsa-miR-373-3p|hsa-miR-520a-3p|hsa-miR-520b-3p|hsa-miR-520c-3p|hsa-miR-520d-3p|hsa-miR-520e-3p                                                                                                                                                                                                                          \n",
      "   19 45410254   C   G   APOE       5       7 None                                                                                                                               hsa-miR-24-3p|hsa-miR-2467-5p|hsa-miR-3934-3p|hsa-miR-4514|hsa-miR-4692                                                                                                                     hsa-miR-1827|hsa-miR-3929|hsa-miR-4478|hsa-miR-4649-3p|hsa-miR-485-5p|hsa-miR-6884-5p|hsa-miR-7160-5p\n",
      "   19 45411774   G   T   APOE      11       1 None                                            hsa-miR-1249-5p|hsa-miR-1321|hsa-miR-4510|hsa-miR-4739|hsa-miR-4756-5p|hsa-miR-6127|hsa-miR-6129|hsa-miR-6130|hsa-miR-6133|hsa-miR-6797-5p|hsa-miR-6873-5p                                                                                                                                                                                                              hsa-miR-3141\n",
      "   19 45403216   C   T TOMM40       3       9 None                                                                                                                                                        hsa-miR-3928-3p|hsa-miR-7106-5p|hsa-miR-766-5p                                                                            hsa-miR-1249-5p|hsa-miR-149-3p|hsa-miR-3126-5p|hsa-miR-4728-5p|hsa-miR-6785-5p|hsa-miR-6797-5p|hsa-miR-6875-5p|hsa-miR-6883-5p|hsa-miR-7845-5p\n",
      "   19 45399051   G   A TOMM40       7       5 None                                                                                            hsa-miR-4651|hsa-miR-608|hsa-miR-6748-5p|hsa-miR-6756-5p|hsa-miR-6766-5p|hsa-miR-6769a-5p|hsa-miR-6769b-5p                                                                                                                                                 hsa-miR-3180|hsa-miR-3180-3p|hsa-miR-3196|hsa-miR-6784-5p|hsa-miR-6816-5p\n",
      "   19 45409090   C   T   APOE       2      10 None                                                                                                                                                                          hsa-miR-5571-3p|hsa-miR-9500                                                                   hsa-miR-1343-3p|hsa-miR-2110|hsa-miR-4271|hsa-miR-4667-5p|hsa-miR-4700-5p|hsa-miR-4725-3p|hsa-miR-6780b-5p|hsa-miR-6783-3p|hsa-miR-6852-5p|hsa-miR-8089\n",
      "   19 45396434   C   G TOMM40       8       3 None                                                                                             hsa-miR-3660|hsa-miR-4288|hsa-miR-4526|hsa-miR-6071|hsa-miR-628-5p|hsa-miR-632|hsa-miR-93-3p|hsa-miR-9718                                                                                                                                                                                  hsa-miR-1231|hsa-miR-4299|hsa-miR-542-3p\n",
      "   19 45404883   A   C TOMM40       2       9 None                                                                                                                                                                          hsa-miR-4648|hsa-miR-6802-5p                                                                                       hsa-miR-1244|hsa-miR-199a-3p|hsa-miR-199b-3p|hsa-miR-3122|hsa-miR-3129-5p|hsa-miR-3913-5p|hsa-miR-4259|hsa-miR-4425|hsa-miR-4703-3p\n",
      "   19 45403005   C   T TOMM40       2       9 None                                                                                                                                                                            hsa-miR-12131|hsa-miR-4487                                                                                              hsa-miR-4446-3p|hsa-miR-4492|hsa-miR-4498|hsa-miR-4505|hsa-miR-4721|hsa-miR-5001-5p|hsa-miR-5787|hsa-miR-6515-5p|hsa-miR-762\n",
      "   19 45401270   G   C TOMM40       9       2 None                                                                          hsa-miR-1587|hsa-miR-3620-5p|hsa-miR-4492|hsa-miR-4498|hsa-miR-4505|hsa-miR-4731-5p|hsa-miR-5001-5p|hsa-miR-5787|hsa-miR-762                                                                                                                                                                                             hsa-miR-10400-3p|hsa-miR-4674\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# -------- CONFIG --------\n",
    "DATA_DIR = \"data\"\n",
    "VCF_PATH = os.path.join(DATA_DIR, \"ALL.chr19.phase3.vcf.gz\")\n",
    "REF_FASTA = os.path.join(DATA_DIR, \"chr19.fa\")\n",
    "GTF_PATH = os.path.join(DATA_DIR, \"gencode.v19.annotation.gtf\")\n",
    "MATURE_FA = os.path.join(DATA_DIR, \"mature.fa\")\n",
    "\n",
    "# genes of interest on chr19 (APOE region)\n",
    "GENES = [\"APOE\", \"TOMM40\"]  # i am focusing on APOE/TOMM40 locus on chr19\n",
    "# window around variant to check for seed matches (bp)\n",
    "WINDOW = 30\n",
    "\n",
    "# -------- helpers --------\n",
    "def parse_gtf_get_utrs(gtf_path, genes_of_interest):\n",
    "    \"\"\"\n",
    "    Returns dict: gene -> list of (start, end, strand) of 3' UTRs (1-based inclusive).\n",
    "    GTF is assumed to be GENCODE v19 format (hg19).\n",
    "    \"\"\"\n",
    "    gene_utrs = {g: [] for g in genes_of_interest}\n",
    "    with open(gtf_path, \"rt\") as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            parts = line.strip().split(\"\\t\")\n",
    "            if len(parts) < 9:\n",
    "                continue\n",
    "            chrom, src, feature, start, end, score, strand, frame, attr = parts\n",
    "            # if feature != \"three_prime_utr\":\n",
    "            #     continue\n",
    "            # parse gene_name from attributes\n",
    "            attrs = {}\n",
    "            for kv in attr.strip().split(\";\"):\n",
    "                kv = kv.strip()\n",
    "                if kv == \"\":\n",
    "                    continue\n",
    "                if \" \" in kv:\n",
    "                    k, v = kv.split(\" \", 1)\n",
    "                    attrs[k] = v.strip().strip('\"')\n",
    "            gene_name = attrs.get(\"gene_name\") or attrs.get(\"gene_id\")\n",
    "            if gene_name in gene_utrs and chrom in (\"19\", \"chr19\"):\n",
    "                s = int(start)\n",
    "                e = int(end)\n",
    "                gene_utrs[gene_name].append((s, e, strand))\n",
    "    return gene_utrs\n",
    "\n",
    "def build_interval_tree(intervals):\n",
    "    it = IntervalTree()\n",
    "    for (s,e,_) in intervals:\n",
    "        # convert to half-open [s, e) for IntervalTree\n",
    "        it[s-1:e] = True\n",
    "    return it\n",
    "\n",
    "def load_miRNA_seeds(mature_fa_path, species_tag=\"hsa\"):\n",
    "    \"\"\"\n",
    "    Load mature.fa, extract human miRNAs (header contains species), produce seed (nt 2-8).\n",
    "    Returns dict: miRNA_name -> seed (7mer sequence, 5'->3').\n",
    "    \"\"\"\n",
    "    seeds = {}\n",
    "    for rec in SeqIO.parse(mature_fa_path, \"fasta\"):\n",
    "        # header like: >hsa-let-7a-5p MI0000060 ...\n",
    "        header = rec.id\n",
    "        if header.startswith(\"hsa-\") or header.startswith(\"hsa\"):\n",
    "            name = rec.id\n",
    "            seq = str(rec.seq).upper()\n",
    "            if len(seq) >= 8:\n",
    "                # canonical seed positions 2-8 (1-based)\n",
    "                seed = seq[1:8]  # python indexing [1:8] => positions 2..8 inclusive\n",
    "                seeds[name] = seed\n",
    "    return seeds\n",
    "\n",
    "def revcomp(seq):\n",
    "    return str(Seq(seq).reverse_complement())\n",
    "\n",
    "def sequence_window_from_fasta(chrom_seq, pos1_based, window=30):\n",
    "    \"\"\"\n",
    "    chrom_seq: the chromosome sequence string (1-based indexing assumed)\n",
    "    pos1_based: variant position (1-based)\n",
    "    returns sequence of length=window*2+1 centered on pos\n",
    "    \"\"\"\n",
    "    L = len(chrom_seq)\n",
    "    start = max(1, pos1_based - window)\n",
    "    end = min(L, pos1_based + window)\n",
    "    # convert to 0-based for python slicing\n",
    "    return chrom_seq[start-1:end]\n",
    "\n",
    "# -------- main --------\n",
    "def main():\n",
    "    # ----1) parse GTF --> get 3'UTR intervals for genes of interest\n",
    "    print(\"Parsing GTF for 3'UTR coordinates (genes: {})...\".format(\", \".join(GENES)))\n",
    "    gene_utrs = parse_gtf_get_utrs(GTF_PATH, GENES)\n",
    "    for g in GENES:\n",
    "        print(f\"  {g}: {len(gene_utrs.get(g,[]))} 3'UTR segments\")\n",
    "\n",
    "    # build combined interval tree\n",
    "    combined_intervals = []\n",
    "    for g, segs in gene_utrs.items():\n",
    "        for (s,e,strand) in segs:\n",
    "            combined_intervals.append((s,e,g,strand))\n",
    "    if not combined_intervals:\n",
    "        print(\"No 3'UTR intervals found for genes. Exiting.\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    tree = IntervalTree()\n",
    "    for s,e,g,strand in combined_intervals:\n",
    "        tree[s-1:e] = (g, strand)  # store gene and strand\n",
    "\n",
    "    # ----2) load chr19 reference sequence\n",
    "    print(\"Loading chr19 reference FASTA...\")\n",
    "    chrom_record = None\n",
    "    for rec in SeqIO.parse(REF_FASTA, \"fasta\"):\n",
    "        # fasta header might be \">chr19\" or \"19\"\n",
    "        if rec.id in (\"19\",\"chr19\",\"NC_000019.9\"):\n",
    "            chrom_record = rec\n",
    "            break\n",
    "        # try contains 'chr19'\n",
    "        if \"chr19\" in rec.id:\n",
    "            chrom_record = rec\n",
    "            break\n",
    "    if chrom_record is None:\n",
    "        # fallback: take the first record (should be chr19)\n",
    "        chrom_record = next(SeqIO.parse(REF_FASTA, \"fasta\"))\n",
    "    chrom_seq = str(chrom_record.seq).upper()\n",
    "    print(\"Reference length:\", len(chrom_seq))\n",
    "\n",
    "    # ----3) load miRNA seeds (human only)\n",
    "    print(\"Loading miRNA seeds from miRBase mature.fa ...\")\n",
    "    seeds = load_miRNA_seeds(MATURE_FA)\n",
    "    print(\"Loaded {} human miRNA seeds\".format(len(seeds)))\n",
    "\n",
    "    # Build set of seed reverse complements (the sequence in mRNA that matches miRNA seed)\n",
    "    seed_rc_map = {}  # seed_rc -> list of miRNA names\n",
    "    for name, seed in seeds.items():\n",
    "        rc = revcomp(seed)\n",
    "        seed_rc_map.setdefault(rc, []).append(name)\n",
    "\n",
    "    # ----4) iterate variants in chr19 VCF and check overlap with 3'UTRs\n",
    "    print(\"Scanning VCF for variants overlapping 3'UTRs...\")\n",
    "    vcf = VCF(VCF_PATH)\n",
    "    results = []\n",
    "    # small optimization: we iterate through entire chr19 but only evaluate variants whose pos is in tree\n",
    "    for var in tqdm(vcf):\n",
    "        pos = var.POS  # 1-based\n",
    "        hits = tree[pos-1]\n",
    "        if not hits:\n",
    "            continue\n",
    "        # variant may be multi-allelic; process each alt allele separately\n",
    "        ref = var.REF\n",
    "        alts = var.ALT\n",
    "        info = var.INFO\n",
    "        # get global AF if present (from INFO or compute later)\n",
    "        af = var.INFO.get('AF') if 'AF' in var.INFO else None\n",
    "\n",
    "        for alt in alts:\n",
    "            # build reference window\n",
    "            seq_ref_window = sequence_window_from_fasta(chrom_seq, pos, window=WINDOW)\n",
    "            # create alt sequence window by replacing reference base(s) with alt at center\n",
    "            # determine relative center index in seq_ref_window\n",
    "            left_len = min(WINDOW, pos-1)\n",
    "            center_idx = left_len  # 0-based index inside seq_ref_window corresponding to REF start\n",
    "            # for indels, handle simply by constructing an alt sequence replacement\n",
    "            # Build local sequence as: left + ref + right\n",
    "            left = seq_ref_window[:center_idx]\n",
    "            right = seq_ref_window[center_idx+len(ref):]\n",
    "            alt_seq_window = left + alt + right\n",
    "\n",
    "            # search for seed RC matches in both sequences\n",
    "            gained = []\n",
    "            lost = []\n",
    "            # naive string search: check all rc seeds\n",
    "            for rc_seed, mirnas in seed_rc_map.items():\n",
    "                if rc_seed in seq_ref_window and rc_seed not in alt_seq_window:\n",
    "                    # seed lost\n",
    "                    lost.extend(mirnas)\n",
    "                if rc_seed not in seq_ref_window and rc_seed in alt_seq_window:\n",
    "                    # seed gained\n",
    "                    gained.extend(mirnas)\n",
    "\n",
    "            # deduplicate and keep counts\n",
    "            gained_set = sorted(set(gained))\n",
    "            lost_set = sorted(set(lost))\n",
    "\n",
    "            results.append({\n",
    "                \"CHROM\": var.CHROM,\n",
    "                \"POS\": pos,\n",
    "                \"REF\": ref,\n",
    "                \"ALT\": alt,\n",
    "                \"GENE\": \";\".join(sorted(set([h.data[0] for h in hits]))),\n",
    "                \"STRAND\": \";\".join(sorted(set([h.data[1] for h in hits]))),\n",
    "                \"AF\": af,\n",
    "                \"GAIN_miRNAs\": \"|\".join(gained_set),\n",
    "                \"LOSS_miRNAs\": \"|\".join(lost_set),\n",
    "                \"n_GAIN\": len(gained_set),\n",
    "                \"n_LOSS\": len(lost_set),\n",
    "                \"ID\": var.ID if var.ID else \".\",\n",
    "            })\n",
    "\n",
    "    print(\"Total candidate variant rows:\", len(results))\n",
    "    df = pd.DataFrame(results)\n",
    "    out_csv = \"results_apoe_miRNA_seed_gain_loss.csv\"\n",
    "    df.to_csv(out_csv, index=False)\n",
    "    print(\"Wrote\", out_csv)\n",
    "    # summary: top candidates by (n_GAIN + n_LOSS)\n",
    "    if not df.empty:\n",
    "        df['impact'] = df['n_GAIN'] + df['n_LOSS']\n",
    "        top = df.sort_values(by='impact', ascending=False).head(20)\n",
    "        print(\"\\nTop candidate variants (top 20 by predicted seed gain/loss):\")\n",
    "        print(top[[\"CHROM\",\"POS\",\"REF\",\"ALT\",\"GENE\",\"n_GAIN\",\"n_LOSS\",\"AF\",\"GAIN_miRNAs\",\"LOSS_miRNAs\"]].to_string(index=False))\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
